File: src/controller.py
from player import Player
from utils import clear


class Controller:
    _player: Player

    def __init__(self, player: Player):
        self._player = player

    def _build_prompt(self):
        prompt = '\nWhat will you do?'
        prompt += '\n  q - Quit'
        prompt += '\n  i - Inventory'

        if self._player.room().north():
            prompt += '\n  n - Move North'

        if self._player.room().south():
            prompt += '\n  s - Move South'

        if self._player.room().east():
            prompt += '\n  e - Move East'

        if self._player.room().west():
            prompt += '\n  w - Move West'

        if self._player.room().items():
            prompt += '\n  take <item> - Take a item from this room'

        if self._player.items():
            prompt += '\n  drop <item> - Drop a item in this room'

        prompt += '\n\nCommand: '
        return prompt

    def tick(self):
        command = input(self._build_prompt())
        action, *subjects = command.split(' ', 1)

        try:

            if action == 'n' or action == 's' or action == 'e' or action == 'w':
                self._player.move(action)
                return command

            if action == 'take' or action == 'get':
                self._player.take(subjects[0])
                return command

            if action == 'drop':
                self._player.drop(subjects[0])
                return command

            if action == 'i' or action == 'inventory':
                clear()
                self._player.inventory()
                return command

            if action == 'q' or action == 'quit':
                return command

            raise NotImplementedError
        except:
            input("INVALID COMMAND! Press ENTER key then enter a valid command...")

    def start(self):
        command = "start"
        while (command is not 'q'):
            clear()
            self._player.look()
            command = self.tick()

File: src/controller.ts
import { Player, Room } from './entities';
import { confirm, prompt, invalid } from './utils';

export class Controller {
  constructor(public readonly player: Player) {}

  private buildPrompt() {
    let prompt = 'What will you do?';
    prompt += '\n  q - Quit';
    prompt += '\n  i - Inventory';

    if (this.player.lookNorth()) prompt += '\n  n - Move North';
    if (this.player.lookSouth()) prompt += '\n  s - Move South';
    if (this.player.lookEast()) prompt += '\n  e - Move East';
    if (this.player.lookWest()) prompt += '\n  w - Move West';
    if (this.player.itemSearch()) prompt += '\n  take <item> - Take a item from this room';
    if (this.player.bagSearch()) prompt += '\n  drop <item> - Drop a item in this room';
    prompt += '\n\nCommand: ';

    return prompt;
  }

  private async tick() {
    const response = await prompt(this.buildPrompt());
    const [action, ...subjects] = (response.text as string).split(' ', 2);

    try {
      let room: Room;
      switch (action) {
        case 'north':
        case 'n':
        case 'south':
        case 's':
        case 'east':
        case 'e':
        case 'west':
        case 'w':
          this.player.move(action);
          return response.text;

        case 'get':
        case 'take':
          this.player.take(subjects[0]);
          return response.text;

        case 'drop':
          this.player.drop(subjects[0]);
          return response.text;

        case 'i':
        case 'inventory':
          console.clear();
          console.log(this.player.toString());
          await confirm();
          return response.text;

        case 'q':
          return response.text;

        default:
          throw new Error('INVALID COMMAND!');
      }
    } catch (err) {
      await invalid(err.message);
    }
  }

  public async loop() {
    let command;
    do {
      console.clear();
      console.log(this.player.look().toString());
      command = await this.tick();
    } while (command !== 'q');
  }
}
